Introduce regression test caching

viktormalik opened the pull request:
Until now, regressions tests always needed downloaded kernels in the `kernel/` directory. We used the `tests/regression/kernel_modules` directory for storing the analysed IR and C files but that was intended for manual inspection only.

This refactors regression tests in a way that they can completely run on data stored in this directory. The directory has been renamed to `tests/regression/test_data` as it better reflects its new purpose. See commit message for the first commit for implementation details on how this is done.

In addition, this refactors pytest unit tests (`tests/unit_tests/*.py`) such that they only require one kernel version (3.10.0-957.el7).

Thanks to the above changes, CI can now cache only the `test_data` directory. In case it can be restored from cache, we only download kernel 3.10.0-957.el7 for unit tests, otherwise we download all the kernels and cache the `test_data` directory after the tests are run. The cache key is composed of the LLVM version and a hash of all test spec files. This way, we cache separate test data for each LLVM version and it gets invalidated whenever the set of test cases changes.

Unfortunately, this means that we no longer build all LLVM IR files in every CI run. But since we still have tests for building them in unit tests, I believe that this is a compromise that we should do for the sake of faster CI runs. Also, I haven't found a way to invalidate the caches in case we do a change in the build process so #321 is still an issue.

Overall, this significantly speeds up the CI. With cached `test_data`, CI runs in ~9 minutes.

This resolves #295 but see [comment](https://github.com/diffkemp/diffkemp/issues/295#issuecomment-2206449246) in that issue to learn about alternatives which can be considered. Since the amount of cached data is significantly lower (~100 MB per LLVM version), we should also get rid of #337 once we delete the old cache entries containing entire kernels.

PLukas2018 reviewed (commented):
I tried to review the changes, but somebody else should probably look at this also, because I am not yet much familiar with this part of DiffKemp.

viktormalik commented on tests/regression/mock_source_tree.py:
```diff
@@ -0,0 +1,63 @@
+        llvm_file = os.path.join(self.source_dir, "{}.ll".format(symbol))
+        src_file = os.path.join(self.source_dir, "{}.c".format(symbol))
+
+        if not os.path.exists(llvm_file) and self.real_source_tree is not None:
```
Actually, we can't. I originally had that but some tests failed when trying to find missing definitions for functions which do not have a definition (or it cannot be found). In such a case, the LLVM file was not created (and copied) but the assertion failed when trying to re-run the tests on the cached data (i.e. without the access to the real source tree).

I think that the correct thing to do is just not compile anything, linking missing definitions will fail, and DiffKemp will return "not equal" (which in this case was the expected result of the test).

viktormalik commented on .github/workflows/ci.yml:
```diff
@@ -93,13 +56,34 @@ jobs:
+          matrix.regression-tests &&
+          steps.cache-test-data.outputs.cache-hit == 'true'
+        run:
+          nix develop .#test-kernel-buildenv --command bash -c \
+            "rhel-kernel-get 3.10.0-957.el7 --output-dir kernel"
```
Yeah, it probably doesn't as getting one version is quite fast. Also, I plan to change the unit tests to use some upstream version as those are much quicker to get (since we're getting a tarball directly instead of extracting it from SRPM).

viktormalik commented on tests/regression/syntax_diff_test.py:
```diff
@@ -90,6 +92,26 @@ def test_syntax_diff(task_spec):
     Function differences are displayed using the diff utility and this test
     cannot be currently used to verify them.
```
Good catch, I updated the comment.

viktormalik commented on tests/regression/sysctls_test.py:
```diff
@@ -66,7 +66,7 @@ def task_spec(request):
             continue
         mod_old, mod_new = spec.build_modules_for_function(fun.name)
-        spec.prepare_dir(
+        spec.cache_dir(
```
A very good point. Originally, the intention was to use this to cache source files, too, but I decided to omit that for now. I removed the `cache_dir` method completely which resolves most of the comments in the review.

viktormalik commented on tests/regression/task_spec.py:
```diff
@@ -122,27 +152,29 @@ def new_src_file(self, name=None):
         return self._file_name("new", "c", name)
 
-    def prepare_dir(self, old_module, new_module, old_src, new_src, name=None):
+    def cache_dir(self, old_module, new_module, old_src, new_src, name=None):
```
The method has been removed.

viktormalik commented on tests/regression/task_spec.py:
```diff
@@ -122,27 +152,29 @@ def new_src_file(self, name=None):
-        if not os.path.isdir(self.task_dir):
-            os.mkdir(self.task_dir)
+        if os.path.isdir(self.task_dir):
+            return
```
The method has been removed.

viktormalik commented on tests/regression/syntax_diff_test.py:
```diff
@@ -72,7 +74,7 @@ def task_spec(request):
     for fun in spec.functions:
         mod_old, mod_new = spec.build_modules_for_function(fun)
-        spec.prepare_dir(
+        spec.cache_dir(
```
The method has been removed.

viktormalik commented:
> I tried to review the changes, but somebody else should probably look at this also, because I am not yet much familiar with this part of DiffKemp.

Thanks for the review! I addressed the comments, I also found that we need to rename the task directories in `tests/regression/test_data` as there may be same functions in different task specs (for different kernel versions) so the task name must also contain the spec name, too.

DanielKriz commented on tests/regression/task_spec.py:
```diff
@@ -40,10 +41,43 @@ class TaskSpec:
+        self.new_kernel_dir = os.path.join(kernel_path, spec["new_kernel"])
+        self.old_kernel_source_tree = None
+        self.new_kernel_source_tree = None
+        if os.path.isdir(self.old_kernel_dir):
+            self.old_kernel_source_tree = KernelSourceTree(
+                self.old_kernel_dir,
+                KernelLlvmSourceBuilder(self.old_kernel_dir))
+        if os.path.isdir(self.new_kernel_dir):
+            self.new_kernel_source_tree = KernelSourceTree(
+                self.new_kernel_dir,
+                KernelLlvmSourceBuilder(self.new_kernel_dir))
```
I am not sure what is going to happen if kernel directories do not exists. I am not sure if I understand correctly, but it would probably fail around [here](https://github.com/viktormalik/diffkemp_/blob/2c876c431fde337176d411df493ffa17db54d044/tests/regression/mock_source_tree.py#L41), right? Wouldn't it be better to fail around here? It is possible to raise and exception in tests.

viktormalik replied:
I don't think that we can fail here. Non-existing kernel directory is a genuine option, the only problem would be when the kernel directory doesn't exist *and* the files are not cached. In order to fail here, we'd have to detect that the cached files are missing and that's hard to do here as we don't know what files to look for (it depends on the task type).

In the current implementation, the failure will occur in `KernelSourceTree` (or `KernelLlvmSourceBuilder`) when trying to find the LLVM IR module for a given function/module/sysctl.

DanielKriz replied:
I think I understand now, that I have revisited it with clarification. I through that when `self.old_kernel_source_tree` is `None`, then it is going to fail, but there is more logic behind it.

DanielKriz commented:
I have gone through the changes and everything I have found is only nitpicking, but I am also not confident in my knowledge of this part of diffkemp.

DanielKriz commented:
The CI is passing and it look that it works correctly, then I think that it is safe to assume that this pull request is working as expected. And the test data are loaded from the cache. From my side there are no objections.

PLukas2018 reviewed (approved):
It looks good to me too.

FrNecas commented on tests/unit_tests/kernel_llvm_source_builder_test.py:
```diff
@@ -9,25 +9,22 @@
+def test_create_kernel():
     """Creating kernel source builder."""
-    builder = KernelLlvmSourceBuilder(kernel_dir)
+    builder = KernelLlvmSourceBuilder("kernel/linux-3.10.0-957.el7")
```
What is the purpose of this change? As far as I can tell, leaving `kernel_dir` as the argument to `KernelLlvmSourceBuilder` would be the same but we would have one less hardcoded string.

viktormalik replied:
Good catch, fixed.

FrNecas reviewed (approved):
A complicated set of changes, indeed! Conceptually, it looks good to me and it looks like the CI is sped up thanks to the changes. I have just one minor nitpick, otherwise looks good to me. Also, regarding the discussion about alternatives in #295 , I also prefer this approach compared to checking the `test_data` directory into git.
